[Infra UI] Metrics Explorer

simianhacker opened the pull request:
This PR was accidentally merged and reverted. The new PR to replace this is located at https://github.com/elastic/kibana/pull/35846

---



## Summary

This PR closes elastic/kibana#28027 by adding the UI for the Metrics Explorer along with the REST endpoint for the retrieving the data.

![image](https://user-images.githubusercontent.com/41702/56993779-36a84b00-6b52-11e9-9134-b5afd7974f17.png)

#### Group By Everything

![image](https://user-images.githubusercontent.com/41702/56993808-4d4ea200-6b52-11e9-8425-ebeecddaeb29.png)

#### Filter by `host.name` group by `system.network.name`

![image](https://user-images.githubusercontent.com/41702/56993899-82f38b00-6b52-11e9-8df7-a8a46ce19012.png)


#### Dark Mode Compatible

![image](https://user-images.githubusercontent.com/41702/56993708-052f7f80-6b52-11e9-9d85-ec8d00e22214.png)


### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- ~~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~ Follow up issue has been created for documentation: https://github.com/elastic/kibana/issues/35840
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [x] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)








### For maintainers

- ~~This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~
- ~~This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~

### Reviewer Notes

The `react-hooks-testng-library` throws a warning about calling not calling `act` for components that update state. This issue https://github.com/mpeyper/react-hooks-testing-library/issues/14 explians that those errors will disappear once `React@16.8.9` is released. Act doesn't allow async/await updates which makes it difficult to avoid this error with hooks with an immediate async call that updates state with its results. In other words, testing React hooks is not really ready for Prime Time.

The `EuiComboBox` is throwing an error for the `inputRef` not being a valid attribute. This is due to 
https://github.com/elastic/eui/issues/1853. It should disappear as soon as EUI is upgraded in Kibana.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7102/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7247/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7308/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7940/)

spalger commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7951/)

ruflin commented:
Tried to run this locally but I get the following:

![image](https://user-images.githubusercontent.com/244900/55536211-77c74f80-56b9-11e9-8ce9-5dd7e987ca04.png)

Anything additional I need to do except running metricbeat?

exekias commented:
It worked for me. This is looking neat @simianhacker, quite clean and powerful!

I have a few comments on usability:

- As a first instinct I wrote the metric I wanted to graph in the search box, it turns out that's the filter. You have to click on the `+` sign to add one. Personally I would prefer the metric name to be the main input, as filtering is secondary here.

![image](https://user-images.githubusercontent.com/299804/55552479-bec73c00-56dd-11e9-96f4-a8155b6e0122.png)

- I suspect at some point we will want to be able to show several lines in the same graph (instead of a single graph per group). This may clash with having 2 different metrics graphed together, or this may be fine. Just trying to make sure we think about it to avoid removing a feature later.

- I love the `Load more charts` as a good way to show something without shooting yourself in the foot. It leaves the question on what order you use by default (probably hardcoded). Once clicked (many charts spawned), do you foresee any issue when hitting `Refresh`? I guess it will just be slower due to more graphs being requested.

And a few UI glitches (nothing important at this stage):

- Randomizing color when adding a new metric would avoid a step for most users (choosing a new color)
- We may want to limit group names length:

![image](https://user-images.githubusercontent.com/299804/55552380-7c9dfa80-56dd-11e9-94f2-f25fcde24988.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8283/)

simianhacker commented:
@ruflin Make sure to run `yarn kbn bootstrap` and restart Kibana.

simianhacker commented:
>As a first instinct I wrote the metric I wanted to graph in the search box, it turns out that's the filter. You have to click on the + sign to add one. Personally I would prefer the metric name to be the main input, as filtering is secondary here.

Good point, I will move the metrics and group by line above the filter and time range. I think that makes more sense. Maybe on first load (with default settings) we should set the popover on the first metric to open, you are immediately presented with the "metric" edit dialog. If any of the settings are different then the defaults we just display it as it with the assumption that the screen has been linked to.

@makwarth Thoughts?

>I suspect at some point we will want to be able to show several lines in the same graph (instead of a single graph per group). This may clash with having 2 different metrics graphed together, or this may be fine. Just trying to make sure we think about it to avoid removing a feature later.

@exekias You can add up to 5 metrics per chart, do we need to add some instruction OR change the design to make that more clear? I'm planning on adding a switch to indicate a separate axis for a metric when people try to overlay `memory` with something like `load`.

>I love the Load more charts as a good way to show something without shooting yourself in the foot. It leaves the question on what order you use by default (probably hardcoded). Once clicked (many charts spawned), do you foresee any issue when hitting Refresh? I guess it will just be slower due to more graphs being requested.

@exekias I'm not sure what the correct behavior is for this one, currently when you click refresh (or change anything... metrics, time range, filter, etc) it RESETS the results and loads the first 9 charts again with an updated query. I choose this behavior because if someone has 50 charts currently loaded we would need to either fetch all 50 (which could be a problem for some clusters). Also because the query is changing (refresh changes the time range), the resulting matches could change. My conclusion was that safest (and easiest) approach is to just reset the pagination at the sacrifice of the user experience for reloading/updating the query after deep pagination.

One idea is to re-paginate the results by taking the total number of visible results, dividing by 9 and making those serialy fill back in. We would probably need to still display a loading screen while all this happens. This is a non-trivial amount of work and adds complexity to the loading logic, I could see this being an enhancement after the initial release.

@makwarth Thoughts?

>Randomizing color when adding a new metric would avoid a step for most users (choosing a new color)

@exekias  Instead of having unlimited colors with the EUI color picker, I was planning giving the user the choice of  5 colors from the color blind friendly palette provided by EUI. When you add a new metric it will try and pick a color that's not being used by one of the other metrics.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8343/)

simianhacker commented:
I just changed first view behavior to the following view, this is what you will get when you first open the Metrics Explorer OR if you open a link to a default set of options.

![image](https://user-images.githubusercontent.com/41702/55595465-d06a0d00-56f8-11e9-846c-46173485311a.png)

I also changed the color picker to a limited colorblind safe palette. When the user adds a new metric it will randomly pick a color that hasn't been used yet, every new metric should be a different color. I also used an ID for the colors instead of the actual color value so if we decide later that we want to give the user the ability to change the palette we can. OR if we ever want to change the color definitions.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8357/)

makwarth commented:
I really like the simplicity of having just one series per graph. Do we need to have the option of five? They get cluttered quickly anyway at this size. 

I also like the simplicity of having a single aggregation applied to all metrics. The metrics explorer is meant for quick exploration, and not dashboarding, so I think it makes sense to optimize for speed and simplicity - but let me know what you think. 

_If_ we had one global aggregation and just one metric series per graph, it all gets much simpler. No aggregation configuration per metric and same color for all series. The metric field input becomes a simple combobox. Users can quickly enter 2-3 metrics and get them graphed instantly. To change the default aggregation, they can use the dropdown.

As for the load more/refresh scenario, I think the current implementation is a good start. We can iterate on this later.

A few nitpicks:
- "by" is not explicit enough, imo. Can we write "Graph per" or something like that?
- if you graph per `container.id` the grid breaks. Truncate titles?
- x-axis values often hard to read

simianhacker commented:
@makwarth I'm confused by "one series per graph" and the "Users can quickly enter 2-3 metrics and get them graphed instantly." Each metric ends up being a series on the chart. Should we just limit the user to one metric? That would simplify this quite a bit because we could just have a single drop down for the field and a single drop down for the aggregation.

I agree with having 5 series/metrics on a chart becomes un-readable but I think users will want to be able to quickly compare two or three. I can imagine the user would want to do things like this:

![image](https://user-images.githubusercontent.com/41702/55633367-0ea31300-5771-11e9-8941-dd4389a8c8e8.png)

Narrow the filter down to 1 host then create a chart per interface. 

As for the charts, I'm still working polishing those up. I want to use the [Elastic Charts](https://elastic.github.io/elastic-charts/) but I'm not sure they are ready. My todo list for the charts consists of:
- [ ] Investigate elastic-charts
- [ ] Customize the tooltip with the timestamp and field names for the series
- [ ] Auto format the series based on the field name. For example for anything that has the `.pct` suffix will be formatted as `0.0%` or `.bytes` as bytes.
- [ ] Fix chart padding (this will fix numbers getting cut off on the axis)
- [ ] Truncate the chart titles
- [ ] Allow multiple Y axis for charting two metrics with different data types. ie. Memory vs Load. (depends on if we decide to go with multiple metrics)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8440/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8446/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8449/)

elasticmachine commented:
Pinging @elastic/infrastructure-ui

simianhacker commented:
@makwarth I split the aggregations out to a seperate separate pull down and then created a combo box for the metrics. I still have all the metrics on a single chart.  The color is picked from a static list based on the order of the selected metrics. I think we should limit the number of metrics to 3, that will support the use case of comparing 2 network metrics (in/out) or 3 the three load metrics.  

![image](https://user-images.githubusercontent.com/41702/55742113-c3dd0180-59e3-11e9-8dcd-731d63129f9a.png)

After talking with @roncohen we interpreted your comment to be: 1 metric per chart, if grouping by a field there will be multiple charts per group by. Personally I prefer to put all the metrics on 1 chart.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8719/)

simianhacker commented:
Chart tooltip:

![image](https://user-images.githubusercontent.com/41702/55755667-7de46580-5a04-11e9-8f37-6c1c04fd59e0.png)

makwarth commented:
Great progress @simianhacker! I love that the metric input is so clean and fast now.

I'm good with the three metric limit. From a UX perspective, it's a bid odd with an arbitrary limit, but I don't think it's a big deal. However, when typing the fourth metric, can we disable the auto-suggestion altogether, so that users understand that they've reached a limit of sorts?

A few other nitpicks:
- I'd love to start with the metric field. Can we do <metric combobox> <agg dropdown> <group combobox>
- Can agg. default to average?
- Can we make the agg dropdown and the group combobox narrower, so there's room for at least three metrics in the metric combobox (at least if they have normal name lengths)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8916/)

simianhacker commented:
@makwarth As far as starting with a metric, I would suggest `avg(system.cpu.user.pct)`. One potential pitfall is that value is really not quite right if there is more then one `host.name`. I would recommend making the initial state have `avg` pre-selected and open the drop down for selecting the first metric,  here is what that initial screen looks like now. 

![image](https://user-images.githubusercontent.com/41702/55841479-2a494900-5ae4-11e9-889c-8fc732e4a066.png)

I also tweaked the "Group By" combobox to be smaller than the "Metrics" box.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8951/)

makwarth commented:
(Updated my comment as the `<>` was cut)

exekias commented:
Just tested this and I quite like the new behavior, right to the point!

makwarth commented:
Regarding the aggregation dropdown and rate, what do you think about having a "rate" checkbox that applies to all metrics? That way the dropdown doesn't need to offer aggs with/without rate. 
Order: `<metric combobox> <rate checkbox> <agg dropdown> <group combobox>`

Re the metric limit: Are you concerned about performance, if we bump the metric limit to e.g. 5 or 10? The metric combobox won't handle that many metric labels very well, and neither will the graphs (y-axis, colors, size) so we might still choose to restrict the number, but curious to hear your thoughts.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9255/)

simianhacker commented:
@makwarth 

> Regarding the aggregation dropdown and rate, what do you think about having a "rate" checkbox that applies to all metrics? That way the dropdown doesn't need to offer aggs with/without rate.
Order: <metric combobox> <rate checkbox> <agg dropdown> <group combobox>

My only concern is that most metrics that require a rate should _only_ be run with the `max` aggregation, otherwise the rate calculated will be wrong. I opted to reduce the chance for errors by making it look like a regular aggregation. Are there use cases where you would want to calculate the rate using `avg` or `min`?

> Re the metric limit: Are you concerned about performance, if we bump the metric limit to e.g. 5 or 10? The metric combobox won't handle that many metric labels very well, and neither will the graphs (y-axis, colors, size) so we might still choose to restrict the number, but curious to hear your thoughts.

Performance is not an issue since it's only creating 9 charts at a time (thanks to how we are paginating) and because the most costly part of the aggregation is the `composite terms` and the `date_histogram`. There isn't much of a performance hit for additional metrics once the documents are filter (by the aggregations).

My main argument for restricting the metrics is purely for UX. I think charts with more than 5 lines starts to become less useful with each additional metric due to readability. 

The other issue is axises, right now everything is being put on the same axis so if you try to compare something that is percent based with something like `document count`, which is a raw number, the percent based metric will likely be impossible to read. The solution would be to detect when the different field types are used and create an axis for each field type. I was planning on adding support for multiple axises next. With that in mind, let's say the user adds 5 different field types, they would end up with 4 additional axises, which starts eating into the chart's screen real estate.

makwarth commented:
@simianhacker Ok, sounds good. 

I do believe there are cases for rate/min, so how about we go with:
- no rate checkbox
- dropdown which includes "rate/max", "rate/min" and "rate/avg" (and "avg", "max", "min", "sum")
- dropdown defaults to "avg"

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9397/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9472/)

simianhacker commented:
@makwarth I added the changes above along with adding the ability to click on a title of a grouping and set a filter, sort of like a drilldown.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9478/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9628/)

simianhacker commented:
@makwarth @roncohen The issue I have with `rate of avg` and `rate of min` is that technically it will lead to the wrong values for the customer. Consider this example, let's say you have a network interface that is sending ~10 bytes per second and we are collection every 1s. This is recorded as a counter, every byte that crosses the wire is just added to an integer then sent to ES. 

Now let's imagine we make a chart with 10 second buckets (or each point equals 10 seconds) for the time range, I believe "last 15 minutes" does this. If we were to look at the raw data it might look something like this: 
```
--- First 10 second bucket ---
1000
1010
1022
1030
1040
1054
1060
1073
1081
1090
--- Second 10 second bucket ---
1090
1090
1090
1090
1090
1090
1090
1090
1140
1190
--- Next 10 sec ---
```

When the user wants to calculate "outbound network rate in seconds" what they really want is to take the max value of the first bucket and subtract it from the max value of the second bucket. This will yield a 10 seconds rate that can be down sampled to a 1 second rate (which is technically the avg rate for that 10 second period). In the example above the math looks like:

```
(1190 - 1090) / 10 = 10 per second
```
Now let's say we used the average of the two buckets instead of the max:

```
1105 - 1046 / 10 = 5.9 per second
```

Finally, let's say we used the min of the two buckets instead of the max:

```
1090 - 1000 / 10 = 9 per second
```

Because this is a counter, there will often be multiple records with the same value then all of a sudden traffic transfers, like you see at the end of the second bucket. The more variability you add to the data the larger the differences become in each approach.

Personally, I would remove `rate of min` and `rate of average` because I really can't find a use case where you would prefer that value over `rate of max`. I also think it's safe to just call it `rate` and document how we calculate that.

roncohen commented:
@simianhacker thanks for the relative path fix. 404 error is gone now. 

Rate/min/max: I will summarize our discussion here: I was under the impression we were discussing `avg of rate` and not `rate of avg`. This is the difference between taking the rate first and then averaging over, for instance, hosts or disks and then taking the average of a time bucket and then taking the rate. I agree it's fine to say that the `rate` option means `rate of max` and remove options for `rate of avg`/`rate of min`.

We also discussed:
* Creating more graphs when a user adds another metric, instead of having multiple metrics on the same graph, with all the problems that brings reg. axis.
* Showing a cog wheel on graphs. When user clicks it, they get a menu: 
  * Linking directly to TSVB where user can add more metrics if they want
  * Adds a filter directly for that "group" (drill down mentioned above)

We should also investigate letting users specify multiple things in the "graph per" field. For example, you want to see all your disks across your fleet, but putting in "device" will not work because each host has a `/dev/sda1`, so you need one graph per `host.name, system.filesystem.device_name` or something along those lines.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9939/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10109/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10127/)

simianhacker commented:
Swapped the charts out for [Elastic Charts](https://github.com/elastic/elastic-charts):

![image](https://user-images.githubusercontent.com/41702/56331065-9e689a00-613f-11e9-88ba-7e6df3b20c00.png)

The charts are current throwing an error in Typescript because of this issue: https://github.com/elastic/elastic-charts/issues/177

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10187/)

simianhacker commented:
@makwarth What are your thoughts on the color palette for the metrics? When I swapped out the charts, I re-aligned with the EUI default palette with the first color being the `green-ish` color. All of our other charts start out with the second color in the palette which is the `blue` color (which I prefer). I really liked the first metric being `blue` and the second metric being the `red-ish` color. Because it make a really good contrast between the two metrics. With the EUI default palette, the first 2 colors are `green-ish` and `blue` which, in my opinion, doesn't provide enough contrast and is less useful.

simianhacker commented:
The current code is throwing a warning from EUI related to this issue: https://github.com/elastic/eui/issues/1853

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10289/)

simianhacker commented:
@roncohen @makwarth For multiple group by fields, how do you want to display this in the results? Keep in mind there will be one set of metrics per combination of groupings. With that in mind, let's take the use case of displaying `rate(system.network.out.bytes)` and `rate(system.network.in.bytes)`, then grouping by `system.network.name` and `host.name`. This will return a results for each matching pair. For my laptop it produces 17 different charts, 1 chart with 2 metrics for each interface that the `host.name` has.

For the early prototype, I just set the title of the chart to `{firstGroupBy}, {secondGroupBy}`.  But this becomes an issue with long names because the second group by name will get cut off on small-ish displays. 

How many group bys should we allow? Technically we can support unlimited but screen real estate becomes an issue for both the Group By combo box and the chart titles.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10315/)

simianhacker commented:
I added support for multiple Y axises...

![image](https://user-images.githubusercontent.com/41702/56433013-51421080-6284-11e9-9f93-56e974793f0e.png)

The way it determines the number of axises is based on the format of the field. In the screenshot above, the user has chosen three different fields; `system.cpu.user.pct`, `system.load.1`, and `system.memory.used.pct`.  The format of each field is determined by the suffix, for `system.cpu.user.pct` the suffix `pct` indicates that this is a percent. For any field that doesn't match a known formatting scheme, it defaults to a plain number formatter. 

When the chart is rendered, we set up an axis for each unique format and then assign the series to the axis. This will allow the users to effectively compare different types of values without having to worry about axises or scale.

If you notice the tooltip for the chart, it's only listing one value and the timestamp is not formatted. This is a bug that the Elastic Charts team is aware of and is working on a fix.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10434/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10637/)

simianhacker commented:
Elastic Charts has a fix for the tooltip and multiple axises. https://github.com/elastic/elastic-charts/pull/187

makwarth commented:
>What are your thoughts on the color palette for the metrics?

I agree the contrast was better with the blue/red. Is it an option to overwrite the order and swap red with green in the metrics explorer?

>How many group bys should we allow? Technically we can support unlimited but screen real estate becomes an issue for both the Group By combo box and the chart titles.

Let's start with just one?

>...y-axis...

As mentioned in our discussion, I worry about the complexity here. What do you think about sticking to a single y-axis? Its unit would be determined by the first metrics series, I guess? As you mentioned, if a users graphs counts and percentages, the y-axis will obviously not make sense for both series, but I don't think that's a big issue for the main use case of this feature (super quick (single) metrics exploration. 
Later, users will be able to click a graph in the metrics explorer and jump to a pre-populated visualize/TSVB graph where they can configure multiple y-axis and store as saved object for a dashboard.

roncohen commented:
for exploration, i think being able to add more than one "group by" is important. You want to be able to check out all the disks of your db servers for example. That would require you to group by host.name, disk.name. Screen real estate: agreed that this will be a problem. I think pagination here is a fine compromise. Again, this is about exploration, people will see a page of graphs and then use the filter to filter down for the ones that are interesting.

> For the early prototype, I just set the title of the chart to {firstGroupBy}, {secondGroupBy}. But this becomes an issue with long names because the second group by name will get cut off on small-ish displays.

I think this is fine. 2 is the highest number of group-bys i can think of that will be useful right now, but i don't see a reason to "artificially" limit it for this first iteration.

roncohen commented:
agree when it comes to multiple y-axis'. We should keep it at one and either graph each metric in its own graph or add them to the same graph and then just use the y-axis that fits with the first metric that was added

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11085/)

simianhacker commented:
Removed support for multiple axises, change the color palette back to blue/red for the first two metrics. Cleaned up the charts.

![image](https://user-images.githubusercontent.com/41702/56698113-b5f6d400-66a5-11e9-9172-8e2176950a3c.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11092/)

simianhacker commented:
Retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11106/)

makwarth commented:
Looks great @simianhacker !

simianhacker commented:
Here is the pull down:

![image](https://user-images.githubusercontent.com/41702/56776948-c5e4e580-6783-11e9-845b-80ef26120d95.png)

And when you click "Open in Visualize" you get this:

![image](https://user-images.githubusercontent.com/41702/56776973-deed9680-6783-11e9-936b-f95bd7294f02.png)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11294/)

makwarth commented:
Wonderful! 👏

roncohen commented:
looks good Chris. Are you planning to add the possibility to group by multiple fields?

btw. group by column now ends with `keyword`. Maybe something i did?
![Screen Shot 2019-04-29 at 5 01 08 PM](https://user-images.githubusercontent.com/744/56905415-70b91480-6aa0-11e9-9f80-20f4839b5331.png)

simianhacker commented:
@roncohen I thought we agreed to defer the multiple group by? I'm going to spend the next few days getting this PR ready for review by Thursday morning (EU time).

As far as `.keyword`, that's because the `metricbeat-*` index template wasn't install before indexing.

roncohen commented:
On Mon, 29 Apr 2019 at 17.45, Chris Cowan <notifications@github.com> wrote:

> @roncohen <https://github.com/roncohen> I thought we agreed to defer the
> multiple group by?
>

Ok. We can always add more 👍

I'm going to spend the next few days getting this PR ready for review by
> Thursday morning (EU time).
>
> As far as .keyword, that's because the metricbeat-* index template wasn't
> install before indexing.
>

Ah, of course. Thanks!

—
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/kibana/pull/34019#issuecomment-487632675>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAAF2H5OPAKCHYKXV6VA3DPS4JZLANCNFSM4HB3VIDQ>
> .
>

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11663/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11687/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11699/)

makwarth commented:
Polish:

1. When clicking the Metrics explorer tab, the copy says "Chart not available". Could this say something encouraging, like "Choose metrics to plot" ?

1. Can we remove the auto-suggestion dropdown once selecting a metric? (Keep cursor active in input field)

1. Should we give items (e.g. `host.name`) in the "graph per" input a label, like the metrics? Probably a grey label?

1. Can we copy the APM design for the Actions popover?
```
ACTIONS
[Infra icon] Add filter
[Visualize icon] Open in Visualize [link out arrow]
```

simianhacker commented:
> When clicking the Metrics explorer tab, the copy says "Chart not available". Could this say something encouraging, like "Choose metrics to plot" ?

Right now we are using EUI's empty chart component for that message. We would need to build a different component. I think we should hold off and make that change in the next release in the intrest of making FF.

> Can we remove the auto-suggestion dropdown once selecting a metric? (Keep cursor active in input field)

Unfortunately, this is the default behavior of the `EuiComboBox`. We will need to file an issue with the @elastic/eui-design team.

> Should we give items (e.g. host.name) in the "graph per" input a label, like the metrics? Probably a grey label?

Done

> Can we copy the APM design for the Actions popover?

We are not going to be able to add the "popout" icon next to the link for "Open in Visualize" because `EuiContextMenuPanelItemDescriptor` doesn't support it. Looks like APM wrote their own component using the componets that make up `EuiContextMenu`. Unfortunately at this point we don't have time to re-write that and get it in for FF. @elastic/eui-design team is coming up with a solution, once that's in we can start using it. Here is how it looks now.

![image](https://user-images.githubusercontent.com/41702/56981955-fa1a2680-6b34-11e9-82dc-415043e87b75.png)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11860/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11866/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11881/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11903/)

simianhacker commented:
I have no idea how this just got merged!

makwarth commented:
@simianhacker we can't change the copy of the EUI empty chart component? "Chart not available" sounds like the app is broken, which is really not what we want to convey to users that see this app for the first time.

simianhacker commented:
@makwarth The on text that’s configurable is the text bellow. I can put together a follow up PR for a replacement component today but I wanted to get the main PR in so the rest of the team has time to review it before FF.

makwarth commented:
++

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12134/)
